### PR TITLE
Virtualize occlusion and transform feedback queries

### LINKED ISSUE #459: Rewrite query implementation to use virtual queries
This is required to gracefully handle cases where queries cross the boundary of render passes.
Currently we either have two options:
- Just emit queries inside render passes and pray app is sensible (breaks stuff)
- Suspend render passes between every query (bad for performance)

### PR DESCRIPTION
**Pull request by doitsujin**
Allows us to allocate more than one Vulkan query per D3D12 query for certain query types, which makes the query optimization work even if a single query spreads across multiple render pass instances. This allows us to drop the workaround for AC:Valhalla.

Closes #459-

### CODE REVIEW

File: libs/vkd3d/shaders/cs_resolve_query.comp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  if (thread_id >= query_count)
+    return;
+
+  // The query map is an array of linked lists, with the
+  // first query_count entries guaranteed to be list heads
+  query_map_entry_t entry = map.entries[thread_id + entry_offset];
+  uint64_t dst_data[c_field_count];
+
+  // By copyig the first query we get the reset for free
```
Comment (HansKristian-Work) [inline libs/vkd3d/shaders/cs_resolve_query.comp:47]: Typo: copyig

File: libs/vkd3d/command.c [hunk 2]
```diff
@@ -2572,31 +2603,44 @@ static void d3d12_command_list_end_active_query(struct d3d12_command_list *list,
...
+    query->heap = heap;
     query->index = index;
-    query->flags = flags;
+    query->type = type;
     query->state = VKD3D_ACTIVE_QUERY_RESET;
+    query->vk_pool = heap->vk_query_pool;
+    query->vk_index = index;
+    return true;
+
+/*    return d3d12_command_allocator_allocate_query(list->allocator,
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c]: I guess this code should be removed?

Comment (HansKristian-Work) [reply libs/vkd3d/command.c]: It's reinstated in a later commit so I guess it makes sense to do this.

File: libs/vkd3d/command.c [hunk 3]
```diff
@@ -2692,6 +2696,339 @@ static void d3d12_command_list_handle_active_queries(struct d3d12_command_list *
...
+        dst_queries += d->virtual_query_count;
+    }
+
+    /* Upload query lists in chunks since vkCmdUpdateBuffer is limited to
+     * 64kiB per invocation. Normally, one single iteration should suffice. */
+    for (i = 0; i < list->pending_queries_count; i += 2048)
+    {
+        unsigned int count = min(2048, list->pending_queries_count - i);
+
+        VK_CALL(vkCmdUpdateBuffer(list->vk_command_buffer, entry_buffer.buffer,
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:2943]: Isn't the scratch buffer host visible?

File: tests/d3d12.c [hunk 4]
```diff
@@ -47492,6 +47492,139 @@ static void test_undefined_read_typed_buffer_as_untyped_dxil(void)
...
+        0x43425844, 0x91af6cd0, 0x7e884502, 0xcede4f54, 0x6f2c9326, 0x00000001, 0x000000b0, 0x00000003,
+        0x0000002c, 0x0000003c, 0x00000070, 0x4e475349, 0x00000008, 0x00000000, 0x00000008, 0x4e47534f,
+        0x0000002c, 0x00000001, 0x00000008, 0x00000020, 0x00000000, 0x00000000, 0x00000003, 0xffffffff,
+        0x00000e01, 0x445f5653, 0x68747065, 0xababab00, 0x52444853, 0x00000038, 0x00000040, 0x0000000e,
+        0x04000059, 0x00208e46, 0x00000000, 0x00000001, 0x02000065, 0x0000c001, 0x05000036, 0x0000c001,
+        0x0020800a, 0x00000000, 0x00000000, 0x0100003e,
+    };
+    static const D3D12_SHADER_BYTECODE ps = {ps_code, sizeof(ps_code)};
+    static const uint32_t expected_results[] = {1,0,1,1,614400,0,307200,307200};
+    static const float depthOne = 1.0f;
```
Comment (HansKristian-Work) [inline tests/d3d12.c:47530]: weird camel case

File: libs/vkd3d/command.c [hunk 5]
```diff
@@ -7433,13 +7513,25 @@ static void STDMETHODCALLTYPE d3d12_command_list_ResolveQueryData(d3d12_command_
...
-                1, &copy_region));
+            VK_CALL(vkCmdCopyBuffer(list->vk_command_buffer,
+                    query_heap->vk_buffer, buffer->vk_buffer,
+                    1, &copy_region));
+        }
+        else
+        {
+            uint32_t dst_index = aligned_dst_buffer_offset / sizeof(uint64_t);
+
+            d3d12_command_list_resolve_binary_occlusion_queries(list,
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:7534]: Shouldn't this be handled in the resolve shader with a spec const bool? I figure it would just write 1 or 0 depending on the final counter value.

File: libs/vkd3d/command.c [hunk 6]
```diff
@@ -2692,6 +2696,339 @@ static void d3d12_command_list_handle_active_queries(struct d3d12_command_list *
...
+    struct dispatch_entry
+    {
+        struct d3d12_query_heap *heap;
+        uint32_t virtual_query_count;
+        uint32_t unique_query_count;
+        VkDeviceSize resolve_buffer_offset;
+        VkDeviceSize resolve_buffer_size;
+    };
+    
+    struct dispatch_entry *dispatches = NULL;
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:2764]: Any way to avoid having to allocate fresh arrays every time?

**Review by HansKristian-Work — COMMENTED** [review]
Some comments. Overall the implementation makes sense, albeit fairly complicated resolve, but not much we can do about that.

File: libs/vkd3d/command.c (hunk 6 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:2764]: Could technically store the arrays in command allocator, but I can see why it gets very hairy.

File: libs/vkd3d/command.c [hunk 7]
```diff
@@ -2692,6 +2696,339 @@ static void d3d12_command_list_handle_active_queries(struct d3d12_command_list *
...
+    {
+        ERR("Failed to allocate query map.\n");
+        goto cleanup;
+    }
+
+    /* Active list for the current dispatch */
+    src_queries = list->pending_queries;
+    dst_queries = query_list;
+
+    for (i = 0; i < dispatch_count; i++)
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:2897]: Quite hard to follow this loop, but I understand the data structure it's trying to emit, so it's probably fine.

File: libs/vkd3d/command.c (hunk 5 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:7534]: We can't do that since you can allocate both binary and non-binary queries from an occlusion query heap.

There's the option to do this per query during the resolve by passing extra data to the shader, but there's a lot of room for the app to fuck up by doing this:
```
cmd->BeginQuery(..., D3D11_QUERY_TYPE_OCCLUSION);
...
cmd->EndQuery(..., D3D11_QUERY_TYPE_OCCLUSION);
cmd->ResolveQueryData(..., D3D11_QUERY_TYPE_BINARY_OCCLUSION);
```
in which case we should still do a binary resolve, even if it was specified incorrectly for the queries.

File: libs/vkd3d/command.c (hunk 6 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:2764]: Can do, it's just annoying, and we shouldn't hit this more than once per frame anyway.

File: libs/vkd3d/command.c (hunk 3 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:2943]: No, scratch buffers are device local.

File: tests/d3d12.c (hunk 4 above)
Comment (doitsujin) [reply tests/d3d12.c:47530]: fixed.

File: libs/vkd3d/shaders/cs_resolve_query.comp (hunk 1 above)
Comment (doitsujin) [reply libs/vkd3d/shaders/cs_resolve_query.comp:47]: fixed.

File: libs/vkd3d/command.c (hunk 3 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:2943]: Of course we can extend the scratch buffer system to allow host-visible buffers as well, but that's a larger change that would require a fair bit of work. Also we'd be doing random access on host memory in that case, which is probably bad.

Comment (HansKristian-Work) [reply libs/vkd3d/command.c:2943]: Right.

File: libs/vkd3d/command.c (hunk 5 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:7534]: Ah, ok. I didn't realize app could resolve with BINARY separately.

File: libs/vkd3d/command.c (hunk 6 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:2764]: Could just mark this as TODO at least so that we have a potential optimization avenue if games are dumb.

**Review by HansKristian-Work — APPROVED** [review]
LGTM, nice work!

File: libs/vkd3d/command.c (hunk 6 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:2766]: Done.
